### ISSUE TITLE
u_mul_e is not deterministic

### ISSUE THREAD

**Issue by zjost** [dmlc/dgl#2478]
## 🐛 Bug

If I run the same graph/data through `u_mul_e` many times, I can get different results even though none of the underlying tensors are changing.  I have only been able to observe this on a particular graph and it unfortunately does not have the same behavior on any of the built-in graphs I have tried.

## To Reproduce
This is the code I am using, which is an implementation of Label Propagation using edge weights:

```python
import torch as th 
from torch import nn
import dgl.function as fn
import dgl

def test_graph_equality(g1, g2):
    assert th.equal(g1.edges()[0], g2.edges()[0])
    assert th.equal(g1.edges()[1], g2.edges()[1])
    assert th.equal(g1.nodes(), g2.nodes())
    
    # Adjacency matrix
    assert th.equal(g1.adj().coalesce().indices(), g2.adj().coalesce().indices())
    assert th.equal(g1.adj().coalesce().values(), g2.adj().coalesce().values())
    assert g1.adj().coalesce().size() == g2.adj().coalesce().size()

    
class LPAMinimal(nn.Module):

    def __init__(self):
        super(LPAMinimal, self).__init__()
        
    def forward(self, graph, y, k_iters):
        g_copy = dgl.graph(graph.edges()) # Just for testing that nothing is being modified
        
        y_out = y.clone().detach().requires_grad_(True)
        y_out_copy = y.clone().detach().requires_grad_(True) # Again, for testing that nothing is changing
        
        ys = list()
        with graph.local_scope():
            for k in range(k_iters):
                # Each iteration should generate the *same* output
                graph.ndata['h'] = y_out
                graph.update_all(
                    fn.u_mul_e('h', 'edge_weight', 'm'), # This causes the problem, even when "edge_weights" are all 1.
                    #fn.copy_u('h', 'm'),  # This makes the problem go away.
                    fn.mean('m', 'h'),
                )
                h = graph.ndata['h']
                ys.append(h)
                
                # test_graph_equality(g_copy, graph) # This passes 
                # assert th.equal(y_out, y_out_copy) # This passes 

                # y_out = h  # Un-commenting this would implement Label Propagation; Ignore for this bug.
                           
        return ys

def lpa_test_small(g, labels):
    lpa = LPAMinimal()
                      
    g.edata['edge_weight'] = th.ones((g.num_edges(), 1))  # Set to 1.0 so should match "fn.copy_u('h', 'm')
    ys = list()
    for epoch in range(10):
        ys += lpa(g.clone(), labels.unsqueeze(1).clone(), k_iters=3) # runs inner loop 3 times
        
    equal = [th.equal(ys[0], ys[i]) for i in range(1, len(ys))] # Tests if all iterations are equal
    return all(equal)
```

## Expected behavior

Since the graph, edge weight tensor, and initial embeddings "y" are all constant, I expect any number of iterations to give identical results.  This is the case for all the built-in dgl.data graphs I have tried, but fails for my graph.  All of the graph equality tests pass and there's no indication that any underlying data structure involved in the computation is changing.

## Environment

 - DGL Version (e.g., 1.0): 0.5.2
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):  PyTorch 1.5.0
 - OS (e.g., Linux):  Amazon Linux (SageMaker notebook, p3 instance)
 - How you installed DGL (`conda`, `pip`, source): pip
 - Python version: 3.6.10
 - CUDA/cuDNN version (if applicable): 10.0
 - GPU models and configuration (e.g. V100): p3 instance
 
## Additional context

It's fairly difficult to get the inconsistency to happen.  It will seem to be consistent for many iterations before generating a new result and can sometimes go back to generating the original, consistent result.  It's been tricky to track down because it's relatively infrequent.  Unfortunately, I can't share my graph publicly.  But reach out to jostza@ in AWS if you need to discuss more details of the data.

**Comment by zjost** [comment]
Update:
If I replace `fn.u_mul_e('h', 'edge_weight', 'm')` by 'lambda e: {'m' : e.src['h'] * e.data['edge_weight']}`, it seems to solve the problem and give consistent results.

**Comment by yzh119** [comment]
@zjost could you please also provide us with the graph you use? I cannot reproduce your bug with a random graph:
```python
g = dgl.rand_graph(100, 300)
y = th.rand(100, 40)
print(lpa_test_small(g, y))
```
and the result is:
```python
[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
```
which is the desired behavior IMO.

**Comment by yzh119** [comment]
The `u_mul_e` should be deterministic since 0.5.0, can you confirm that you were using dgl 0.5.2 by:
```python
print(dgl.__version__)
```

**Comment by zjost** [comment]
I uninstalled and re-installed now using version 0.5.3 and confirm the problem still exists.  As I mentioned above, I can't share this graph publicly, but I work at AWS, so please reach out to me at "jostza@" on Slack and we can figure something out.

**Comment by changym3** [comment]
Hey! I had a similar problem about `u_mul_e`. I want to train a large graph by sampling, but in some blocks, the built-in `fn.u_mul_e` will consitently produce `NaN` values. However, the `lambda e: {'m' : e.src['h'] * e.data['edge_weight']}` works fine. 

But it's confusing that when I use `torch.save()` or `dill.dump()` to serialize this block into disk, and reload it to reproduce the bug. The built-in `fn.u_mul_e` works exactly right.

Is something wrong about my serialization? 

 **Environment**
- GL Version (e.g., 1.0): 0.5.1
- Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.6.0
- OS (e.g., Linux): Ubuntu 18.04 LTS
- How you installed DGL (conda, pip, source): pip
- Python version: 3.6.10
- CUDA/cuDNN version (if applicable): 10.2 / cudnn7.6.5
- GPU models and configuration (e.g. V100): RTX 2080 Ti
'

**Comment by yzh119** [comment]
Thanks @changym3 , I'll have a try,

**Comment by yzh119** [comment]
@changym3 the reason you encounter this issue is that the node feature has dtype float64 while the edge feature has dtype float32. DGL did not auto cast their data type to the same, and the cuda kernel encounter in-alignment issue, resulting in NaN values. You can either use float32 node features (that's the recommended way because it's faster and saves GPU memory) or convert edge feature to float64.

@zjost please check if it's also your cases.

I'll create a patch to check if input's data types are matched. Thanks very much for your feedback.
